File: Backend_Api/src/main/java/com/euroroute/controller/NotificationController.java
package com.euroroute.controller;

import com.euroroute.dto.NotificationDTO;
import com.euroroute.service.NotificationService;
import com.euroroute.repository.UserRepository;
import com.euroroute.entity.User;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/notifications")
@RequiredArgsConstructor
@Slf4j
public class NotificationController {

    private final NotificationService notificationService;
    private final UserRepository userRepository;

    /**
     * Get active notifications for the authenticated user
     */
    @GetMapping("/active")
    public ResponseEntity<List<NotificationDTO>> getActiveNotifications() {
        String userId = getCurrentUserId();
        return ResponseEntity.ok(notificationService.getActiveNotifications(userId));
    }

    /**
     * Get all notifications for the authenticated user
     */
    @GetMapping
    public ResponseEntity<List<NotificationDTO>> getAllNotifications() {
        String userId = getCurrentUserId();
        return ResponseEntity.ok(notificationService.getAllNotifications(userId));
    }

    /**
     * Get unread notifications count
     */
    @GetMapping("/unread/count")
    public ResponseEntity<Map<String, Long>> getUnreadCount() {
        String userId = getCurrentUserId();
        long count = notificationService.getUnreadNotificationsCount(userId);
        return ResponseEntity.ok(Map.of("count", count));
    }

    /**
     * Mark a notification as read
     */
    @PostMapping("/{notificationId}/read")
    public ResponseEntity<NotificationDTO> markAsRead(@PathVariable String notificationId) {
        return ResponseEntity.ok(notificationService.markAsRead(notificationId));
    }

    /**
     * Mark all notifications as read
     */
    @PostMapping("/read-all")
    public ResponseEntity<Void> markAllAsRead() {
        String userId = getCurrentUserId();
        notificationService.markAllAsRead(userId);
        return ResponseEntity.ok().build();
    }

    /**
     * Dismiss a notification
     */
    @PostMapping("/{notificationId}/dismiss")
    public ResponseEntity<Void> dismissNotification(@PathVariable String notificationId) {
        notificationService.dismissNotification(notificationId);
        return ResponseEntity.ok().build();
    }

    /**
     * Delete a notification
     */
    @DeleteMapping("/{notificationId}")
    public ResponseEntity<Void> deleteNotification(@PathVariable String notificationId) {
        notificationService.deleteNotification(notificationId);
        return ResponseEntity.ok().build();
    }

    /**
     * Helper method to get current user ID from security context
     */
    private String getCurrentUserId() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication == null || !authentication.isAuthenticated()) {
            throw new RuntimeException("User not authenticated");
        }

        String userEmail = authentication.getName();
        log.info("Looking up user with email: {}", userEmail);

        // Find user by email and return their UUID
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new RuntimeException("User not found: " + userEmail));

        log.info("Found user: {} with ID: {}", userEmail, user.getId());
        return user.getId();
    }
}

File: euro-route-mobile/lib/shared/controllers/notification_controller.dart
import 'package:get/get.dart';

import '../models/user_model.dart';
import '../services/notification_service.dart';

class NotificationController extends GetxController {
  late final NotificationService _notificationService;

  final notifications = <Notification>[].obs;
  final unreadCount = 0.obs;
  final isLoading = false.obs;

  @override
  void onInit() {
    super.onInit();
    _notificationService = Get.find<NotificationService>();

    // Listen to notification service changes
    everAll(
        [_notificationService.notifications, _notificationService.unreadCount],
        (_) {
      notifications.value = _notificationService.notifications;
      unreadCount.value = _notificationService.unreadCount.value;
    });

    loadNotifications();
  }

  Future<void> loadNotifications({bool unreadOnly = false}) async {
    isLoading.value = true;
    try {
      await _notificationService.getNotifications(unreadOnly: unreadOnly);
      notifications.value = _notificationService.notifications;
      unreadCount.value = _notificationService.unreadCount.value;
    } catch (e) {
      Get.snackbar(
        'Error',
        'Failed to load notifications: $e',
        snackPosition: SnackPosition.BOTTOM,
      );
    } finally {
      isLoading.value = false;
    }
  }

  Future<void> markAsRead(String notificationId) async {
    try {
      await _notificationService.markAsRead(notificationId);
    } catch (e) {
      Get.snackbar(
        'Error',
        'Failed to mark notification as read: $e',
        snackPosition: SnackPosition.BOTTOM,
      );
    }
  }

  Future<void> markAllAsRead() async {
    try {
      await _notificationService.markAllAsRead();
    } catch (e) {
      Get.snackbar(
        'Error',
        'Failed to mark all as read: $e',
        snackPosition: SnackPosition.BOTTOM,
      );
    }
  }

  Future<void> dismissNotification(String notificationId) async {
    try {
      await _notificationService.dismissNotification(notificationId);
    } catch (e) {
      Get.snackbar(
        'Error',
        'Failed to dismiss notification: $e',
        snackPosition: SnackPosition.BOTTOM,
      );
    }
  }

  Future<void> deleteNotification(String notificationId) async {
    try {
      await _notificationService.deleteNotification(notificationId);
    } catch (e) {
      Get.snackbar(
        'Error',
        'Failed to delete notification: $e',
        snackPosition: SnackPosition.BOTTOM,
      );
    }
  }
}
